Cubical/Algebra/Matrix/RowTransformation.agda

```diff
@@ -0,0 +1,770 @@
+
+    transRowsIndRel : (M : FinMatrix A (suc m) n) → (i : Fin m) → Rel (M (suc i)) (transRows M (suc i))
+    transRowsIndRel {m = suc m} M zero    = indRel _
+    transRowsIndRel {m = suc m} M (suc i) = transRowsIndRel _ i
+
+  module _
+    (Rel3 : FinVec A n → FinVec A n → FinVec A n → Type ℓ)
+    (invRow  : (M : FinMatrix A 2 n) → M zero ≡ T M zero)
+    (indRel3 : (M : FinMatrix A 2 n) → Rel3 (M zero) (M one) (T M one)) where
+
+    transRowsIndRel3 : (M : FinMatrix A (suc m) n) → (i : Fin m) → Rel3 (M zero) (M (suc i)) (transRows M (suc i))
+    transRowsIndRel3 {m = suc m} M zero    = indRel3 _
+    transRowsIndRel3 {m = suc m} M (suc i) =
+      subst (λ V → Rel3 V (M (suc (suc i)))
+        (transRows M (suc (suc i)))) (sym (invRow _)) (transRowsIndRel3 _ i)
+
+  module _
+    (Rel : FinVec A n → FinVec A n → Type ℓ)
+    (indRel : (M : FinMatrix A 2 n) → Rel (M zero) (M one) → M zero ≡ T M zero) where
+
+    transRowsIndRelInv :
+        (M : FinMatrix A (suc m) n)
+      → ((i : Fin m) → Rel (M zero) (M (suc i)))
+      →  M zero ≡ transRows M zero
+    transRowsIndRelInv {m = 0}     _ _   = refl
+    transRowsIndRelInv {m = suc m} M rel =
+      let invRow = indRel _ (rel zero)
+          rel₁ = (λ i → subst (λ V → Rel V (M (suc (suc i)))) invRow (rel (suc i)))
+      in  invRow ∙ transRowsIndRelInv _ rel₁
+
+  module _
+    (P : FinVec A n → Type ℓ)
+    (Rel : FinVec A n → FinVec A n → Type ℓ)
+    (indPRel : (M : FinMatrix A 2 n) → P (M zero) → Rel (M zero) (M one) → M zero ≡ T M zero) where
+
+    transRowsIndPRelInv :
+        (M : FinMatrix A (suc m) n)
+      → P (M zero)
+      → ((i : Fin m) → Rel (M zero) (M (suc i)))
+      → M zero ≡ transRows M zero
+    transRowsIndPRelInv {m = 0}     _ _ _   = refl
+    transRowsIndPRelInv {m = suc m} M p rel =
+      let invRow = indPRel _ p (rel zero)
+          p₁   = subst P invRow p
+          rel₁ = (λ i → subst (λ V → Rel V (M (suc (suc i)))) invRow (rel (suc i)))
+      in  invRow ∙ transRowsIndPRelInv _ p₁ rel₁
+
+  module _
+    (P : FinVec A n → Type ℓ)
+    (indP₀ : (M : FinMatrix A 2 n) → P (M zero) → P (M one) → P (T M zero))
+    (indP₁ : (M : FinMatrix A 2 n) → P (M zero) → P (M one) → P (T M one )) where
+
+    transRowsIndP₀ :
+        (M : FinMatrix A (suc m) n)
+      → P (M zero)
+      → ((i : Fin m) → P (M (suc i)))
+      → P (transRows M zero)
+    transRowsIndP₀ {m = 0}     _ p _ = p
+    transRowsIndP₀ {m = suc m} _ p q = transRowsIndP₀ _ (indP₀ _ p (q zero)) (q ∘ suc)
+
+    transRowsIndP₁ :
+        (M : FinMatrix A (suc m) n)
+      → P (M zero)
+      → ((i : Fin m) → P (M (suc i)))
+      →  (i : Fin m) → P (transRows M (suc i))
+    transRowsIndP₁ {m = 0}     _ _ q = q
+    transRowsIndP₁ {m = suc m} _ p q zero    = indP₁ _ p (q zero)
+    transRowsIndP₁ {m = suc m} M p q (suc i) = transRowsIndP₁ _ (indP₀ _ p (q zero)) (q ∘ suc) i
+
+-- Row transformation of linear coefficient matrices
+
+open import Cubical.Algebra.Ring
```

Comment:
Please move import to the top of the file

Reply:
done